src/server/response.rs

```diff
@@ -113,46 +107,206 @@ impl<'a> Response<'a, Fresh> {
         try!(write!(&mut self.body, "{}", self.headers));
         try!(write!(&mut self.body, "{}", LINE_ENDING));
 
-        let stream = if chunked {
-            ChunkedWriter(self.body.into_inner())
-        } else {
-            SizedWriter(self.body.into_inner(), len)
+        Ok(body_type)
+    }
+
+
+}
+
+impl<'a> Response<'a, Fresh> {
+    /// Creates a new Response that can be used to write to a network stream.
+    #[inline]
+    pub fn new(stream: &'a mut (Write + 'a)) -> Response<'a, Fresh> {
+        Response {
+            status: status::StatusCode::Ok,
+            version: version::HttpVersion::Http11,
+            headers: header::Headers::new(),
+            body: ThroughWriter(stream),
+            _writing: PhantomData,
+        }
+    }
+
+    /// Consume this Response<Fresh>, writing the Headers and Status and creating a Response<Streaming>
+    pub fn start(mut self) -> io::Result<Response<'a, Streaming>> {
+        let body_type = try!(self.write_head());
+        let (version, body, status, headers) = self.deconstruct();
+        let stream = match body_type {
+            Body::Chunked => ChunkedWriter(body.into_inner()),
+            Body::Sized(len) => SizedWriter(body.into_inner(), len)
         };
 
         // "copy" to change the phantom type
         Ok(Response {
-            version: self.version,
+            version: version,
             body: stream,
-            status: self.status,
-            headers: self.headers,
-            _marker: PhantomData,
+            status: status,
+            headers: headers,
+            _writing: PhantomData,
         })
     }
-
     /// Get a mutable reference to the status.
     #[inline]
     pub fn status_mut(&mut self) -> &mut status::StatusCode { &mut self.status }
 
     /// Get a mutable reference to the Headers.
+    #[inline]
     pub fn headers_mut(&mut self) -> &mut header::Headers { &mut self.headers }
 }
 
+
 impl<'a> Response<'a, Streaming> {
     /// Flushes all writing of a response to the client.
+    #[inline]
     pub fn end(self) -> io::Result<()> {
-        debug!("ending");
-        try!(self.body.end());
+        trace!("ending");
+        let (_, body, _, _) = self.deconstruct();
+        try!(body.end());
         Ok(())
     }
 }
 
 impl<'a> Write for Response<'a, Streaming> {
+    #[inline]
     fn write(&mut self, msg: &[u8]) -> io::Result<usize> {
         debug!("write {:?} bytes", msg.len());
         self.body.write(msg)
     }
 
+    #[inline]
     fn flush(&mut self) -> io::Result<()> {
         self.body.flush()
     }
 }
+
+#[derive(PartialEq)]
+enum Body {
+    Chunked,
+    Sized(u64),
+}
+
+type StatusFresh = PhantomData<Fresh>;
+
+impl<'a, T: Any> Drop for Response<'a, T> {
+    fn drop(&mut self) {
+        if (&self._writing as &Any).is::<StatusFresh>() {
```

Comment:
That said we can get better perf by just doing `TypeId::of::<W>() == TypeId::of::<StatusFresh>()`, which will get the branch optimized away because monomorphization.